### PR TITLE
Adding benchmark for MoeFlatmmKernel

### PR DESCRIPTION
**Pull request by amd-wsung102**
## Motivation

Previously there was no benchmarking script for `MoeFlatmmKernel` under `/aiter-test/op_tests/op_benchmarks`. This PR adds such a script to make it easier to obtain the performance of `MoeFlatmmKernel`.

## Features

This benchmark features several flags to customize its behavior.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/wsung102/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/wsung102/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link=blue vlink=purple>

Flag | Description
-- | --
--num-iters | Number of timed iterations
--num-warmup | Number of warmup iterations
--seed | Random seed for reproducibility
--case-index | Run one shape case by index (0-based). Default: -1 (run all)
--csv | Optional CSV output path for benchmark results

</body>

</html>

## Benchmark Result
This benchmark is expected to produce output in the following format using the `--csv` flag. This specific output was obtained on MI350 using tensor shapes in GPT-OSS 120B, but it has been tested on MI355 as well. Stage 1 refers to up+gate projection, and stage 2 refers to down projection.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/wsung102/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/wsung102/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link=blue vlink=purple>

block_m | topk | expert | stage1_m | stage1_n | stage1_k | stage1_us | stage1_tflops | stage2_m | stage2_n | stage2_k | stage2_us | stage2_tflops | total_us | num_iters | num_warmup
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
64 | 4 | 128 | 16384 | 6144 | 3072 | 3391.137 | 182.3799 | 65536 | 3072 | 3072 | 1804.361 | 685.534 | 5195.498 | 10 | 5
16 | 4 | 128 | 256 | 6144 | 3072 | 206.1254 | 46.8825 | 1024 | 3072 | 3072 | 118.9647 | 162.463 | 325.0901 | 10 | 5
16 | 4 | 128 | 128 | 6144 | 3072 | 206.8422 | 23.36002 | 512 | 3072 | 3072 | 121.6364 | 79.44721 | 328.4787 | 10 | 5
16 | 4 | 128 | 64 | 6144 | 3072 | 182.6196 | 13.22925 | 256 | 3072 | 3072 | 106.3207 | 45.4459 | 288.9402 | 10 | 5
16 | 4 | 128 | 48 | 6144 | 3072 | 167.0149 | 10.84897 | 192 | 3072 | 3072 | 97.76044 | 37.06897 | 264.7753 | 10 | 5
16 | 4 | 128 | 32 | 6144 | 3072 | 129.352 | 9.338546 | 128 | 3072 | 3072 | 77.489 | 31.17757 | 206.841 | 10 | 5
16 | 4 | 128 | 16 | 6144 | 3072 | 90.95589 | 6.640359 | 64 | 3072 | 3072 | 59.59978 | 20.26785 | 150.5557 | 10 | 5
16 | 4 | 128 | 8 | 6144 | 3072 | 63.10644 | 4.785405 | 32 | 3072 | 3072 | 35.36822 | 17.07691 | 98.47467 | 10 | 5
16 | 4 | 128 | 4 | 6144 | 3072 | 33.59056 | 4.495161 | 16 | 3072 | 3072 | 24.937 | 12.11011 | 58.52756 | 10 | 5
16 | 4 | 128 | 2 | 6144 | 3072 | 20.84367 | 3.622082 | 8 | 3072 | 3072 | 20.36811 | 7.413301 | 41.21178 | 10 | 5
16 | 4 | 128 | 1 | 6144 | 3072 | 16.817 | 2.244677 | 4 | 3072 | 3072 | 19.28811 | 3.914197 | 36.10511 | 10 | 5
32 | 4 | 128 | 8192 | 6144 | 3072 | 1640.756 | 188.4727 | 32768 | 3072 | 3072 | 940.4934 | 657.6072 | 2581.249 | 10 | 5

</body>

</html>

##

### CODE REVIEW

File: op_tests/op_benchmarks/hip/bench_moe_flatmm.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    w2_qt = shuffle_weight_a16w4(w2_qt, 16, False)
+    w1_scale = shuffle_scale_a16w4(w1_scale, case.expert, True)
+    w2_scale = shuffle_scale_a16w4(w2_scale, case.expert, False)
+
+    # per_1x32 cktile kernels dereference exp_bias optional in generated code.
+    bias1 = torch.zeros((case.expert, case.stage1_n), dtype=dtypes.fp32, device=device)
+    bias2 = torch.zeros((case.expert, model_dim), dtype=dtypes.fp32, device=device)
+
+    stage1_out = torch.empty((token, case.topk, inter_dim), dtype=dtype, device=device)
+    _, stage1_us = run_perftest(
```
Comment (ChuanLi1101) [inline op_tests/op_benchmarks/hip/bench_moe_flatmm.py:181]: cktile_moe_stage1 is called twice: once inside run_perftest (which runs num_iters + num_warmup times), and then again explicitly to produce stage2_in. After run_perftest finishes, stage1_out already contains valid output from the last iteration. The second call wastes GPU time, especially for large shapes like M=16384.

**Review by ChuanLi1101 — DISMISSED** [review]
I think it looks overall good. Left some comments for your reference. However, should pass CI test.

**Review by ChuanLi1101 — COMMENTED** [review]
Please fix the CI test failure.

File: op_tests/op_benchmarks/hip/bench_moe_flatmm.py (hunk 1 above)
Comment (amd-wsung102) [reply op_tests/op_benchmarks/hip/bench_moe_flatmm.py:181]: Fixed in this commit https://github.com/ROCm/aiter/pull/2154/changes/99be3165ba7f189a3c9c9f3ba2e449307de726ef. cktile_moe_stage1 is now called once only and stage2_in uses stage1_out.

**Comment by carlushuang** [comment]
@gyohuangxin for new CI flow

**Comment by zufayu** [comment]
Hi, thank you for your contribution!

Over the past six months, ROCm/aiter has undergone significant updates and improvements. We're closing this PR for now to keep the backlog manageable — if you're still able to reproduce the issue or the feature is still relevant, we'd love to hear from you! Please give the latest version of the repo a try and feel free to reopen this PR or submit a new one if the problem persists. Looking forward to your continued contributions!
